로그인, 프로필 화면 [STEP 2] brendan, hayden

cms7371 opened the pull request:
# 구현 내용
### ProfileViewScene
- 프로필 화면을 구현하기 위해 스토리보드에 구성한 scene입니다. 이번 목표가 코드를 통해 화면을 구성하는 것이었기에 만들기만 했습니다.
### ProfileViewController
내부 뷰를 배치하기 위해 구현한 컨트롤러입니다.
```swift
lazy var lowerButtonStackView: UIStackView = {
    let stackView = UIStackView()
    stackView.translatesAutoresizingMaskIntoConstraints = false
    stackView.alignment = .center
    stackView.axis = .horizontal
    stackView.distribution = .fillEqually
    return stackView
  }()
```
브랜던의 아이디어로 화면을 구성할 뷰들을 lazy 멤버로 선언하여 관리 및 활용이 용이하도록 했습니다.
클래스 멤버로 만들어서 생성된 뷰를 어디서든 관계를 맺어줄 수 있고,
lazy로 선언하는 것을 통해 컨트롤러가 호출되는 시점이 아닌 뷰가 호출되는 시점에 초기화 될 수 있도록 했습니다

```swift
private func layoutButtonLabels() {
    view.addSubview(messageLabel)
    view.addSubview(phoneLabel)
    view.addSubview(quoteLabel)
    
    let labelButtonPairs: [(UILabel, UIButton)] = [(messageLabel, messageButton),
                                                     (phoneLabel, phoneButton),
                                                     (quoteLabel, quoteButton)]
    
    labelButtonPairs.forEach { pair in
        let (label, button) = pair
        label.translatesAutoresizingMaskIntoConstraints = false
        label.centerXAnchor.constraint(equalTo: button.centerXAnchor).isActive = true
        label.topAnchor.constraint(equalTo: lowerButtonStackView.bottomAnchor, constant: 15).isActive = true
    }
}
```
그렇게 생성된 뷰들을 연관된 것들끼리 배치하는 함수를 만들어 배치했습니다.
이때 같은 형태로 반복적으로 배치되는 것들을 고차함수로 최대한 묶어봤습니다.

```swift
override func viewDidLoad() {
    super.viewDidLoad()
}

override func viewDidLayoutSubviews() {
    super.viewDidLayoutSubviews()
    layoutBackgroundImageView()
    layoutCloseButton(withSize: 22)
    layoutUpperButtonStackView(withSize: 26)
    layoutLowerButtonStackView()
    layoutButtonLabels()
    layoutSeparatorView()
    layoutProfileInfo()
    }
```
마지막으로 이렇게 구성한 함수들을 호출하여 뷰들이 최종적으로 배치될 수 있도록 했습니다.
didLoad가 아닌 didloadSubviews에서 이를 호출한 것은
```swift
profileImageView.layer.cornerRadius = profileImageView.bounds.width * 0.3
profileImageView.clipsToBounds = true
```
프로필 이미지를 비율에 맞춰 둥글게 해주는 과정에서 뷰의 width를 얻어야하는데
didLoad에서 배치를 하게 될 경우 width가 0으로 나오는 문제가 발생했고, 이를 해결하기 위해 didLoadSubview에서 호출하게 되었습니다
# 고민/궁금한 점
- 코드로 레이아웃을 구성하는 경우 빌드해서 직접 테스트해보는 것 이외에 구성을 미리 볼 수 있는 방법이 없을까요?? 지금이야 작은 프로젝트라 만들고 앱을 실행해보고, 하면서 점검했지만, 앱의 크기가 커지면 이것이 거의 불가능할 것이라 생각했습니다
- didLoad에서 뷰의 크기를 구할 수 있는 방법이 있을까요? 또는 그런 경우가 필요할 때는 didLoadSubviews에서 뷰를 생성하는 것이 맞는 방법인 것인가요? 이 부분은 저희가 더 공부해봐야 할 것 같은데 왜 didLoad에서 뷰를 호출하는 것이 베이스고, 다른 사이클에서 이를 호출하면 어떤 일이 일어나는지 궁금했습니다!

항상 꼼꼼하게 리뷰해주시는 코다 정말 감사드립니다!!!

cc.) @dacodaco @gjwlsdnr0115

gjwlsdnr0115 committed:
feat: Step2 closeButton 구현

- view에 추가
- constraints 구성

gjwlsdnr0115 committed:
feat: Step2 프로필 화면 구현

- 프로필 화면 구현
- 버튼 프로필 이미지, 배경화면 asset 추가

dacodaco commented on LoginProfile/Assets.xcassets/backgroundImg.imageset/Contents.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "images" : [
```
배경이 멋있네요...

dacodaco commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,246 @@
+
+class ProfileViewController: UIViewController {
+
+    lazy var backgroundImageView: UIImageView = {
+        let imageView = UIImageView()
+        let backgroundImage = UIImage(named: "backgroundImg")
+        imageView.image = backgroundImage
+        imageView.contentMode = .scaleAspectFill
+        return imageView
+    }()
```
lazy 키워드를 사용해보셨네요!
lazy 키워드를 UIComponent에 사용할 때 얻을 수 있는 이점에는 무엇이 있고, 유의해야할 점에는 무엇이 있을까요?
현재 보이고 있는 View들은 대부분 ViewController가 나타날 때부터, 사라질 때 까지 모두 함께하는 View인데요!
초기화 시점을 달리해줌으로서 어떤 점이 이득이 될 수 있을지 궁금합니다.

cms7371 replied:
이번 경우에는 화면의 구성과 함께 모든 뷰가 나타나게 되어 lazy가 큰 의미가 없지만,
나중에 동적인 화면을 구성하게 되는 경우 나중에 나타나는 뷰들이 미리 초기화가 되는 것에서 메모리 낭비가 발생할 수 있기 때문에
lazy 키워드를 사용하여 불러지는 시점에 초기화되는 것이 이점이 있다고 생각합니다.
유의해야할 점은 lazy 키워드가 thread-safe하지 않다는 것인데 UI를 구성하는 것은 무조건 메인 스레드에서 이뤄져야하기 때문에 괜찮다고 판단했습니다.

dacodaco commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,246 @@
+        return imageView
+    }()
+            
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
```
View의 계층 구조를 만들어주는 메서드 (`addSubview`) 는 `LayoutSubviews` 메서드에서 진행해주는 것이 좋을까요, 혹은 `viewDidLoad` 에서 진행해주는 것이 좋을까요?

cms7371 replied:
아주아주 좋은 지적 감사합니다!!
알아보기 위해서 각 라이프 사이클에 대해 자세히 찾아봤는데 layoutSubviews는 화면이 회전되거나 변화하는 경우에도 다시 호출된다는 걸 알게되었습니다.
그래서 테스트해보니 layoutSubview에서 화면을 구성하기 위한 것들을 호출하게 되면 여러번 호출된다는 문제점이 발생했습니다
다행히 addSubview가 말썽을 일으키진 않았지만, constraint가 여러번 중복되어 정의되는 문제점을 찾을 수 있었습니다.
그래서 viewDidLoad에서 호출하는 것으로 바꾸고, 프로필 이미지의 코너를 바꿔주는 부분만 willLayoutSubview에서 호출하는 방향으로 리팩터를 진행해보려합니다.

dacodaco replied:
제가 이야기 하고자 하는 부분을 완벽하게 짚어주셨네요...!

dacodaco commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,246 @@
+    
+    private func layoutBackgroundImageView() {
+        view.addSubview(backgroundImageView)
+        view.sendSubviewToBack(backgroundImageView)
```
`sendSubviewToBack` 메서드를 호출한 이유가 궁금합니다!
이미 subView 계층에서 가장 밑에 있지 않은가요?!

gjwlsdnr0115 replied:
코드상 backgroundImageView를 추가하는 부분이 먼저 실행되서 큰 상관은 없지만, 혹시 이후 누군가가 코드의 순서를 바꾸거나 리팩토링 과정에서 안전하게 하기 위해 sendSubviewToBack 메서드를 사용했습니다.

dacodaco replied:
View가 앞으로도 subView의 마지막에 온다는 확신이 있다면, 이런 방법도 좋은 것 같네요 :)

dacodaco commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,246 @@
+    }
+    
+    private func layoutCloseButton(withSize size: CGFloat) {
+        view.addSubview(closeButton)
```
현재 layout 과 관련한 메서드 내부에서 addSubview를 호출해주고 있네요!
`addSubview` 메서드를 layout 관련 메서드 내부에서 호출한 이유가 있을까요?

cms7371 replied:
저희도 코멘트를 보고 다시 생각해보니 적절하지 않다고 판단되어 분리하여 리팩토링 진행해보겠습니다!

dacodaco commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,246 @@
+        view.addSubview(backgroundImageView)
+        view.sendSubviewToBack(backgroundImageView)
+        
+        backgroundImageView.translatesAutoresizingMaskIntoConstraints = false
+        backgroundImageView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
+        backgroundImageView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
+        backgroundImageView.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
+        backgroundImageView.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
```
Constraints 관련 메서드 역시 Layout을 호출하는 메서드에서 진행해주는 이유가 궁금합니다..!!

cms7371 replied:
constraints를 정해주는 것 자체가 레이아웃을 잡아주는 것이라 생각해 하나로 묶었습니다

dacodaco replied:
constraints와 레이아웃은 엄연히 다른 부분이라고 생각해요.
constraints를 바탕으로 레이아웃이 설정되지만, 세팅해주는 순서는 constraints가 더 이를 수 있다고 생각하거든요..!

dacodaco commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,246 @@
+        profileImageView.widthAnchor.constraint(equalTo: safeArea.widthAnchor, multiplier: 0.25).isActive = true
+        profileImageView.heightAnchor.constraint(equalTo: profileImageView.widthAnchor).isActive = true
+        
+        profileImageView.layer.cornerRadius = profileImageView.bounds.width * 0.3
```
`profileImageView` 의 크기는 각 View의 Layout이 결정되기 전에는 알 수 없습니다.
viewDidLoad가 호출된 시점에는 rootView와 subviews 의 레이아웃이 결정되기 이전이므로, 아직 크기가 결정되기 전이라고 볼 수 있을 것 같습니다.
따라서 모든 뷰의 layout이 결정된 이후에 호출해주어야 반영되는 것으로 보입니다.
---
만약 constraints를 viewDidLoad에서 설정하고, cornerRadius를 viewDidLayoutSubview 에서 설정해준다면 어떤 결과가 나올까요?

dacodaco commented:
PR 잘 받았습니다! 전체적으로 가독성 높은 코드로 한번에 구조를 파악할 수 있어서,
코드 리뷰를 하기에도 정말 좋았습니다! 수고 많으셨어요.
---
고민한 부분에 대해서 제 의견을 이야기해볼까 합니다. 
1. 현업에서는 별도의 디버깅 툴을 활용해 확인해볼 수 있는 것으로 알고 있습니다...만 제가 아직 사용해본 적이 없어서 정확한 답변을 드리기가 어렵네요.. 사실 어느정도 익숙해지면, 특정 View의 Constraints가 어디에 위치하는지를 지정해놓기 떄문에, View의 구조를 파악하는 것이 생각보다 어렵지 않습니다. 오히려 각 View를 클릭해야 Constraints를 확인할 수 있는 지금이 더 불편하다고 느껴질지도 모르겠습니다.ㅎㅎ
2. viewDidLoad는 뷰가 준비되는 순서 상 layout이 결정되기 전에 호출됩니다. 즉, 각 뷰의 크기가 결정되기 전에 호출되는 것이겠죠. 따라서 저는 vIew의 constraints 설정은 viewDidLoad에서 진행하고, view의 레이아웃이 설정된 이후 작업해야하는 경우가 있다면 viewDidLayoutSubview 에서 호출하는 방식으로 진행하고 있습니다. 무엇이 되었든 정답은 없지만, Render Loop 라는 키워드로 한번 공부해보시면 좋을 것 같습니다!

cms7371 committed:
refactor: Step 2 프로필 화면 리팩터링

- addSubview 호출 분리
- layout구성을 viewDidLoad로 옮김
- 프로필 코너 라운딩 처리 분리 및 viewDidLayoutSubviews에 배치
